[BC-415] Rework CombinedChainDataClient.getStateAtSlot()

mbaxter opened the pull request:
## PR Description
Rework `CombinedChainDataClient.getStateAtSlot()`.

To avoid race conditions, don't query based on head block root.  First try to query from recent data, and then query from historical state to control for case where blocks are being rapidly imported, and the slot being queried is finalized during processing.

Also, added some test utilities to make it easier to write integration tests:  
- Added `MockRocksDbInstance`, which allows us to create an in-memory version of `RocksDbDatabase`
- Added `InMemoryStorageSystem` which sets up and wires together an in-memory database with a `RecentChainData` instance, and makes the `RecentChainData` instance as well as a `CombinedChainData` instance available.
- Added a `ChainUpdater` that generates chain data with `ChainGenerator` and immediately adds the data to a `RecentChainData` instance.  

Fixes https://github.com/PegaSysEng/teku/issues/1713

mbaxter committed:
Rework CombinedChainDataClient.getStateAtSlot()

Don't query based on head block root.  First try to query from recent
data, and then query from historical state to control for case where
blocks are being rapidly imported, and the slot being queried is
finalized during processing.

mbaxter commented on storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java:
```diff
@@ -149,11 +147,6 @@ public CombinedChainDataClient(
   }
 
-  private boolean isHistoricalData(final UnsignedLong slot) {
-    final boolean finalizedPastFirstEpoch = !recentChainData.getFinalizedEpoch().equals(ZERO);
```
I don't see why we need a special case for the first epoch / genesis block, so I cut this.  But if there's a valid reason, I can add this back in.

ajsutton replied:
I believe it was added because we were always going to disk to retrieve the genesis state due to it being inherently finalised.  But I think we've fixed `isFinalized` now so the last finalized slot is expected to be in memory.

Basically as long as the genesis state is retrieved from the memory store instead of disk it's fine to cut it.

mbaxter replied:
Ok, so basically we want to retrieve latest finalized data from memory, whether or not its on the epoch boundary?  And this applies for all finalized data, not just genesis?

mbaxter committed:
Revert store check change

mbaxter committed:
Rework historical data check

mbaxter committed:
Break up ChainBuilder logic

mbaxter committed:
Create in-memory RocksDbDatabase system

mbaxter committed:
Fix datastructures ssz dependencies

mbaxter committed:
Add InMemoryStorageSySystem utility

mbaxter committed:
Add CombinedChainData integration tests

mbaxter committed:
Revert "Break up ChainBuilder logic"

This reverts commit 54230c305655690043978c82399255c3377b18c3.

mbaxter committed:
Fix spotless errors

mbaxter committed:
Remove unit tests duplicated by integration tests

mbaxter commented on storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbDatabase.java:
```diff
@@ -53,7 +53,11 @@
+    return createV3(db, stateStorageMode);
+  }
+
+  static Database createV3(final RocksDbAccessor db, final StateStorageMode stateStorageMode) {
```
Created this `RocksDbAccessor` interface so that we could have a mock in-memory replacement for the core RocksDB access logic.  Not super happy with the current implementation, but leaving a larger refactor for the future since we have active development in this area.

mbaxter committed:
Add more tests, fix bug

Consistently return the latest state in effect when querying from
recent or historical chain data

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java:
```diff
@@ -107,7 +106,7 @@ public CombinedChainDataClient(
       return getBlockByBlockRoot(headBlockRoot);
     }
-    if (isHistoricalData(slot)) {
+    if (!isRecentData(slot)) {
       LOG.trace("Block at slot {} is in a finalized epoch. Retrieving from historical data", slot);
       return historicalChainData.getLatestFinalizedBlockAtSlot(slot);
     }
```
Aren't we at risk of race conditions here?  ie we check if the slot is recent and it is, but then before we actually get the block the store is changed and the slot we're interested in is dropped.  Seems like we should try to get it from the Store and if it's not there, then fall back to getting it from disk.  That would be a reasonable way to handle the case where we need to request a block that's more than `SLOTS_PER_HISTORICAL_ROOT` behind the current state as well - we'd fail to get it from the store and fallback to getting it from disk.

mbaxter replied:
I think this method currently has a few problems - coming back to it in a follow-up PR.

ajsutton commented on storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/schema/Schema.java:
```diff
@@ -35,4 +35,18 @@
+
+  static Stream<RocksDbVariable<?>> streamVariables(Class<? extends Schema> schema) {
+    return Arrays.stream(schema.getDeclaredFields())
+        .filter(f -> (f.getModifiers() & Modifier.STATIC) > 0)
```
```suggestion
        .filter(f -> Modifiers.isState(f.getModifiers()))
```
and could apply the same change above in `streamColumns` because apparently you should use the `java.lang.reflect.Modifiers` class to decode modifiers...

ajsutton reviewed (approved):
LGTM.

mbaxter committed:
Merge branch 'master' into bc-415/rework-get-state-at-slot

mbaxter committed:
Fix static field check
